### PR TITLE
[BACKLOG-35905] - CSRF - Protect the User Impersonation From Trusted IP Using Parameters authentication mechanism

### PR DESCRIPTION
**Pull request by spravasahu**

### CODE REVIEW

**Review by dcleao — CHANGES_REQUESTED** [review]

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml
Comment (spravasahu) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml]: addressed review comments

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java
Comment (dcleao) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java:481]: In this context, the `AccessDeniedException` should be logged and then converted into calling:
```java
response.sendError( HttpStatus.SC_FORBIDDEN );
```

`chain.doFilter(...)` should not be called in this case.

The ServletException and IOException exceptions are being correctly handled — should be thrown to the caller of `doFilter`.

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (dcleao) [inline extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java:57]: A test is missing that tests the case where AccessDeniedException is thrown.

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java
Comment (spravasahu) [reply extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java]: addressed review comments

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java
Comment (spravasahu) [reply extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java]: addressed review comments

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (spravasahu) [reply extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java:63]: addressed review comments

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java
Comment (dcleao) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilter.java:294]: The test for `csrfValidator` being non-null is already done inside `doCsrfValidation`. I think it's best to do it there, not here.

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (dcleao) [inline extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java:311]: The whole try/catch should not be here, including the `fail` and `assertSame` tests.
The `InternalAuthenticationServiceException` is not being thrown (anymore).
The `IOException` is being thrown directly from `doFilter` and from `testWhenCsrfValidationFailsWithGivenExceptionThenRethrows`.
The `@Test( expected = IOException.class )` is the only needed assertion.

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (dcleao) [inline extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java:327]: Idem. See `testWhenCsrfValidationFailsWithIOExceptionThenRethrowsAsInternalAuthenticationServiceException`.

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (dcleao) [inline extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java:362]: Please, use `never()` instead of `times( 0 )`.

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (dcleao) [inline extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java:393]: Typo: `CSrf` -> `Csrf`

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (spravasahu) [reply extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java]: addressed review Comments

File: extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java
Comment (spravasahu) [reply extensions/src/test/java/org/pentaho/platform/web/http/filters/ProxyTrustingFilterTest.java]: addressed review Comments

**Review by dcleao — APPROVED** [review]
Will merge once the security-web dependency is merged.

**Pull request merged**
